tools/generator/Java.Interop.Tools.Generator.CodeGeneration/CodeGenerator.cs

```diff
@@ -847,7 +855,7 @@ public void WriteInterfaceListenerEventsAndProperties (InterfaceGen @interface,
 			var methods = target.Methods.Concat (target.Properties.Where (p => p.Setter != null).Select (p => p.Setter));
 			var props = new HashSet<string> ();
 			var refs = new HashSet<string> ();
-			var eventMethods = methods.Where (m => m.IsListenerConnector && m.EventName != String.Empty && m.ListenerType == @interface).Distinct ();
+			var eventMethods = methods.Where (m => m.IsListenerConnector && m.EventName != String.Empty && m.ListenerType == @interface).GroupBy (m => m.Name).Select (g => g.First()).Distinct ();
```

Comment:
We have some instances where the generator will now try to add 2 of the same event, for example , some interface have a method AddXyzListener(listener) and AddXyzListener(listener, handler)

here are a couple of outputs errors without the GroupBy

  obj/Debug/android-29/mcw/Android.AccessibilityServices.AccessibilityService.cs(875,130): error CS0102: The type 'AccessibilityService.SoftKeyboardController' already contains a definition for 'ShowModeChanged' [/Users/gugavaro/Work/microsoft/Xamarin/xamarin-android-2/src/Mono.Android/Mono.Android.csproj]
  obj/Debug/android-29/mcw/Android.Media.Session.MediaSessionManager.cs(553,103): error CS0102: The type 'MediaSessionManager' already contains a definition for 'Session2TokensChanged' [/Users/gugavaro/Work/microsoft/Xamarin/xamarin-android-2/src/Mono.Android/Mono.Android.csproj]

this is the code generated without the GroupBy:

```
#region "Event implementation for Android.AccessibilityServices.AccessibilityService.SoftKeyboardController.IOnShowModeChangedListener"
			public event EventHandler<Android.AccessibilityServices.AccessibilityService.SoftKeyboardController.ShowModeChangedEventArgs> ShowModeChanged {
				add {
					global::Java.Interop.EventHelper.AddEventHandler<Android.AccessibilityServices.AccessibilityService.SoftKeyboardController.IOnShowModeChangedListener, Android.AccessibilityServices.AccessibilityService.SoftKeyboardController.IOnShowModeChangedListenerImplementor>(
							ref weak_implementor_AddOnShowModeChangedListener,
							__CreateIOnShowModeChangedListenerImplementor,
							AddOnShowModeChangedListener,
							__h => __h.Handler += value);
				}
				remove {
					global::Java.Interop.EventHelper.RemoveEventHandler<Android.AccessibilityServices.AccessibilityService.SoftKeyboardController.IOnShowModeChangedListener, Android.AccessibilityServices.AccessibilityService.SoftKeyboardController.IOnShowModeChangedListenerImplementor>(
							ref weak_implementor_AddOnShowModeChangedListener,
							Android.AccessibilityServices.AccessibilityService.SoftKeyboardController.IOnShowModeChangedListenerImplementor.__IsEmpty,
							__v => RemoveOnShowModeChangedListener (__v),
							__h => __h.Handler -= value);
				}
			}
			public event EventHandler<Android.AccessibilityServices.AccessibilityService.SoftKeyboardController.ShowModeChangedEventArgs> ShowModeChanged {
				add {
					global::Java.Interop.EventHelper.AddEventHandler<Android.AccessibilityServices.AccessibilityService.SoftKeyboardController.IOnShowModeChangedListener, Android.AccessibilityServices.AccessibilityService.SoftKeyboardController.IOnShowModeChangedListenerImplementor>(
							ref weak_implementor_AddOnShowModeChangedListener,
							__CreateIOnShowModeChangedListenerImplementor,
							AddOnShowModeChangedListener,
							__h => __h.Handler += value);
				}
				remove {
					global::Java.Interop.EventHelper.RemoveEventHandler<Android.AccessibilityServices.AccessibilityService.SoftKeyboardController.IOnShowModeChangedListener, Android.AccessibilityServices.AccessibilityService.SoftKeyboardController.IOnShowModeChangedListenerImplementor>(
							ref weak_implementor_AddOnShowModeChangedListener,
							Android.AccessibilityServices.AccessibilityService.SoftKeyboardController.IOnShowModeChangedListenerImplementor.__IsEmpty,
							__v => RemoveOnShowModeChangedListener (__v),
							__h => __h.Handler -= value);
				}
			}
			WeakReference weak_implementor_AddOnShowModeChangedListener;
			Android.AccessibilityServices.AccessibilityService.SoftKeyboardController.IOnShowModeChangedListenerImplementor __CreateIOnShowModeChangedListenerImplementor ()
			{
				return new Android.AccessibilityServices.AccessibilityService.SoftKeyboardController.IOnShowModeChangedListenerImplementor (this);
			}
#endregion

```

the GroupBy will preference the event generation for the first method, the one with less arguments (without the handler argument).


https://developer.android.com/reference/android/accessibilityservice/AccessibilityService.SoftKeyboardController

Reply:
Do we need to do something like `OrderBy (g => g.Parameters.Count).First ()` to ensure we always get the one with the fewest parameters?  If not it seems like "First" will be undefined.

Reply:
thanks a good suggestion and will not hurt to have it. By my tests, the methods with less arguments are always coming first, but "just in case" let's be safer than sorry here. :)